src/coreclr/src/jit/hwintrinsic.cpp

```diff
@@ -528,8 +532,12 @@ GenTree* Compiler::addRangeCheckIfNeeded(NamedIntrinsic intrinsic, GenTree* immO
             )
     {
         assert(!immOp->IsCnsIntOrI());
-        GenTree* upperBoundNode = gtNewIconNode(immUpperBound, TYP_INT);
-        GenTree* index          = nullptr;
+        assert(immLowerBoundIncl < immUpperBoundExcl);
+
+        ssize_t  adjustedUpperBound = (ssize_t)immUpperBoundExcl - (ssize_t)immLowerBoundIncl;
+        GenTree* upperBoundNode     = gtNewIconNode(adjustedUpperBound, TYP_INT);
+        GenTree* index              = nullptr;
+
         if ((immOp->gtFlags & GTF_SIDE_EFFECT) != 0)
```

Comment:
This isn't part of this PR, but shouldn't this be `impCloneExpr` instead (https://github.com/dotnet/runtime/issues/35314)?

Reply:
Probably, I can take a look at this